### PR TITLE
Speed up gradient clipping and allow parameters on multiple devices.

### LINKED ISSUE #2766: Speed up gradient norm clipping.
## 🚀 Feature
Improve the speed of gradient clipping.

### Motivation
I've noticed that clipping the gradients is quite slow, e.g., for a ResNet50 model (with some minor extensions) gradient clipping takes ~207 ms. The major bottlenecks are the `torch.where` computation within a loop and the manual norm computation.

### Pitch
With slight modifications that don't hurt readability and work well on TPU,
grad_norm can be computed in ~75ms, i.e.,  2.7x as fast. For implementation details see the PR I'm going to link in a second.

Reproduce: Train any model you've got lying around with gradient_clip_val=0.1 and take a look at the trace timeline ;)

As a reference, here's a before and after comparison for my model (trace after 40 warmup steps):
Current:
image

With speed improvements:
image

### PR DESCRIPTION
**Pull request by PhilJd**
## What does this PR do?
This PR speeds up gradient clipping.

The speed up is achieved by:
- Moving `torch.where` out of the loop (and replacing with `min` for simplicity).
- Replacing manual sum and pow with `torch.norm`. While this does some
  unnecessary computation (computing pow(root)) this is still a lot faster.
- Pre-allocating the output gives a slight speed up.

Note that calling .to for all parameters results in a small speed
penalty (~4 ms in my case) but allows parameters on different devices
(matching the pytorch implementation).

Overall this reduces the time used for gradient clipping from 206ms to
74 ms for my model (Resnet50 + few additional vars, all vars on GPU).

I've made sure that this works on TPU in a colab.

Fixes #2766 

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests? 
- [X] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

**Comment by williamFalcon** [comment]
awesome!  this is a great addition.

**Comment by PhilJd** [comment]
Wow, that was fast, thanks a lot for merging! :)

**Comment by Borda** [comment]
what is the speed-up about, just see moving the if block left... 🐰

**Comment by PhilJd** [comment]
The if is unrelated, I just switched to return early to decrease the indent for readability. The main speed up comes from replacing [this](https://github.com/PyTorchLightning/pytorch-lightning/pull/2767/files#diff-04c921d5f6e62ca05170862ee5cdd860L87) and [this](https://github.com/PyTorchLightning/pytorch-lightning/pull/2767/files#diff-04c921d5f6e62ca05170862ee5cdd860L79-L83)
